File: algorithms/sorting-algorithms/merge-sort/C/merge-sort.c
#include <stdio.h>
#include <stdlib.h>

void merge_sort(int *, int, int);
void merge(int *, int, int, int);

int main()
{

    int arr[6] = {1, -5, 3, 10, 0, 18};
    merge_sort(arr, 0, 5);

    for (int i = 0; i < 6; ++i)
    {
        printf("%d ", arr[i]);
    }

    return 0;
}

void merge_sort(int *arr, int left, int right)
{
    if (left >= right)
        return;

    int mid = left + (right - left) / 2;
    int size = right - left;

    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);
    merge(arr, left, mid, right);
}

void merge(int *arr, int left, int mid, int right)
{
    int start1 = left;
    int start2 = mid + 1;
    int end1 = mid;
    int end2 = right;
    int k = 0;
    int *output = (int *)malloc((end2 - start1 + 1) * sizeof(int));

    while (start1 <= end1 && start2 <= end2)
    {
        if (arr[start1] <= arr[start2])
        {
            output[k++] = arr[start1++];
        }
        else
        {
            output[k++] = arr[start2++];
        }
    }

    while (start1 <= end1)
    {
        output[k++] = arr[start1++];
    }

    while (start2 <= end2)
    {
        output[k++] = arr[start2++];
    }

    k = 0;
    for (int i = left; i <= right; ++i)
    {
        arr[i] = output[k++];
    }

    free(output);
}

File: algorithms/sorting-algorithms/merge-sort/JS/merge-sort.js
function mergeSort(arr, left, right) {
  if (left >= right) return;

  let mid = left + Math.floor((right - left) / 2);

  mergeSort(arr, left, mid);
  mergeSort(arr, mid + 1, right);
  merge(arr, left, mid, right);
}

function merge(arr, left, mid, right) {
  let start1 = left;
  let start2 = mid + 1;
  let end1 = mid;
  let end2 = right;
  let k = 0;
  let res = [];

  while (start1 <= end1 && start2 <= end2) {
    if (arr[start1] <= arr[start2]) {
      res.push(arr[start1++]);
    } else {
      res.push(arr[start2++]);
    }
  }

  while (start1 <= end1) {
    res.push(arr[start1++]);
  }

  while (start2 <= end2) {
    res.push(arr[start2++]);
  }

  for (let i = left; i <= right; ++i) {
    arr[i] = res[k++];
  }
}

let arr = [5, 4, 3, 2, 1];
mergeSort(arr, 0, 4);
console.log(arr);
